### PR TITLE
Tickets/dm 35952

### PR DESCRIPTION
**Pull request by teweitsai**
* Begin to implement the simulation mode.
* Support the assignment of debug level.

### CODE REVIEW

**Review by pkubanek — APPROVED** [review]
Please see in-code comments for improvements in doc strings, user messages and logging. The code itself is fine and should do what's intended.

File: doc/developer-guide/developer-guide.rst [hunk 1]
```diff
@@ -46,6 +46,7 @@ The model–view–controller (MVC) architecture is used in this module.
 In this design, the view always shows the data sent from the model.
 This helps to minimize the business logic in view and makes the tests easier.
 If you want to cache the data for a smooth showing in view, you need to do this in the **Model**.
+The controller in MVC reuses the existed module in `ts_m2com <https://github.com/lsst-ts/ts_m2com>`_, which provides the mock server for the simulation purpose as well.
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply doc/developer-guide/developer-guide.rst]: Fixed in the commit: [a7be87c](https://github.com/lsst-ts/ts_m2gui/pull/19/commits/a7be87c8d2aea384541dd1e1fc9a0085be3cc715). Thanks!

File: python/lsst/ts/m2gui/application.py [hunk 2]
```diff
@@ -89,14 +89,26 @@ async def main(argv):
     )
     parser.addOption(option_simulation)
 
+    option_log_level = QCommandLineOption(
+        ["l", "level"],
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2gui/application.py]: Fixed in the commit: [a7be87c](https://github.com/lsst-ts/ts_m2gui/pull/19/commits/a7be87c8d2aea384541dd1e1fc9a0085be3cc715). Thanks!

File: python/lsst/ts/m2gui/main_window.py [hunk 3]
```diff
@@ -238,62 +265,139 @@ def _add_tool_bar(self):
...
+            result = await dialog.show()
+
+            if result == QMessageBoxAsync.Ok:
+                # Close the running asynchronous tasks first. This is to deal
+                # with the condition that the asynchronous tasks are running
+                # but there is no connection actually.
+                if self.model.run_loops:
+                    await self.model.close_tasks()
+
+                app = QApplication.instance()
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2gui/main_window.py]: Fixed in the commit: [a7be87c](https://github.com/lsst-ts/ts_m2gui/pull/19/commits/a7be87c8d2aea384541dd1e1fc9a0085be3cc715). Thanks!

File: python/lsst/ts/m2gui/widget/__init__.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+# This file is part of ts_m2gui.
+#
+# Developed for the LSST Data Management System.
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2gui/widget/__init__.py]: Fixed in the commit: [a7be87c](https://github.com/lsst-ts/ts_m2gui/pull/19/commits/a7be87c8d2aea384541dd1e1fc9a0085be3cc715). Thanks!

File: doc/developer-guide/developer-guide.rst [hunk 5]
```diff
@@ -104,6 +105,22 @@ m2gui.display
...
+    :caption: Class diagram of widget module
+
+* **QMessageBoxAsync** is the QT message box in the asynchronous version.
+* **QFileDialogAsync** is the QT file dialog in the asynchronous version.
+
+Some of the functions in QT widgets (such as `QDialog exec <https://doc.qt.io/qt-6/qdialog.html#exec>`_) are not asynchronous and may spin an additional event loop when running the application.
+Therefore, it is needed to have an asynchronous version of these widgets to avoid breaking the
+trick of switching the event loop used in `qasync <https://github.com/CabbageDevelopment/qasync>`_ library.
+Otherwise, the system would not have the permission to wake up the running `asynchronous tasks <https://docs.python.org/3/library/asyncio-task.html#task-object>`_.
+
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply doc/developer-guide/developer-guide.rst:123]: Fixed in the commit: [a7be87c](https://github.com/lsst-ts/ts_m2gui/pull/19/commits/a7be87c8d2aea384541dd1e1fc9a0085be3cc715). Thanks!

File: python/lsst/ts/m2gui/controltab/tab_diagnostics.py [hunk 6]
```diff
@@ -108,17 +109,20 @@ async def _callback_reboot_controller(self):
...
 
-        dialog.setStandardButtons(QMessageBox.Ok | QMessageBox.Cancel)
-        dialog.setDefaultButton(QMessageBox.Cancel)
+        dialog.setStandardButtons(QMessageBoxAsync.Ok | QMessageBoxAsync.Cancel)
+        dialog.setDefaultButton(QMessageBoxAsync.Cancel)
 
-        decision = dialog.exec()
-        if decision == QMessageBox.Ok:
-            run_command(self.model.reboot_controller)
+        # Block the user to interact with other running widgets
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2gui/controltab/tab_diagnostics.py]: Fixed in the commit: [a7be87c](https://github.com/lsst-ts/ts_m2gui/pull/19/commits/a7be87c8d2aea384541dd1e1fc9a0085be3cc715). Thanks!

File: python/lsst/ts/m2gui/main_window.py [hunk 7]
```diff
@@ -238,62 +265,139 @@ def _add_tool_bar(self):
...
         action_settings = tool_bar.addAction("Settings", self._callback_settings)
         action_settings.setToolTip("Show the application settings")
 
     @asyncSlot()
     async def _callback_exit(self):
-        """Exit the application."""
+        """Exit the application.
+
+        The 'exit' action will be disabled first. If cancels, it will be
+        enabled again.
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2gui/main_window.py]: Fixed in the commit: [a7be87c](https://github.com/lsst-ts/ts_m2gui/pull/19/commits/a7be87c8d2aea384541dd1e1fc9a0085be3cc715). Thanks!

File: python/lsst/ts/m2gui/main_window.py [hunk 8]
```diff
@@ -238,62 +265,139 @@ def _add_tool_bar(self):
...
+        """
+
+        action_exit = self._get_action("Exit")
+        action_exit.setEnabled(False)
+
+        if self.model.system_status["isCrioConnected"]:
+            await prompt_dialog_warning(
+                "_callback_exit()", "There is still the connection with M2 controller."
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2gui/main_window.py]: Fixed in the commit: [a7be87c](https://github.com/lsst-ts/ts_m2gui/pull/19/commits/a7be87c8d2aea384541dd1e1fc9a0085be3cc715). Thanks!

File: python/lsst/ts/m2gui/main_window.py [hunk 9]
```diff
@@ -238,62 +265,139 @@ def _add_tool_bar(self):
...
+            )
+
+        else:
+            dialog = self._create_dialog_exit()
+            result = await dialog.show()
+
+            if result == QMessageBoxAsync.Ok:
+                # Close the running asynchronous tasks first. This is to deal
+                # with the condition that the asynchronous tasks are running
+                # but there is no connection actually.
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2gui/main_window.py]: Fixed in the commit: [a7be87c](https://github.com/lsst-ts/ts_m2gui/pull/19/commits/a7be87c8d2aea384541dd1e1fc9a0085be3cc715). Thanks!

File: python/lsst/ts/m2gui/main_window.py [hunk 10]
```diff
@@ -238,62 +265,139 @@ def _add_tool_bar(self):
...
 
-        dialog = QMessageBox()
-        dialog.setIcon(QMessageBox.Warning)
+        dialog = QMessageBoxAsync()
+        dialog.setIcon(QMessageBoxAsync.Warning)
         dialog.setWindowTitle("exit")
 
         if self.model.system_status["isCrioConnected"]:
-            dialog.setText(
-                "There is still the connection with M2 controller. Exit the application?"
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2gui/main_window.py:278]: This is to remove the code.

File: python/lsst/ts/m2gui/main_window.py [hunk 11]
```diff
@@ -238,62 +265,139 @@ def _add_tool_bar(self):
...
 
         return dialog
 
     @asyncSlot()
     async def _callback_connect(self):
-        """Callback function to connect to the M2 controller."""
-        run_command(self.model.connect)
+        """Callback function to connect to the M2 controller.
+
+        The 'connect' action will be disabled first. After the connection, it
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2gui/main_window.py]: Fixed in the commit: [a7be87c](https://github.com/lsst-ts/ts_m2gui/pull/19/commits/a7be87c8d2aea384541dd1e1fc9a0085be3cc715). Thanks!

File: python/lsst/ts/m2gui/main_window.py [hunk 12]
```diff
@@ -238,62 +265,139 @@ def _add_tool_bar(self):
...
+        will be enabled again.
+        """
+
+        action_connect = self._get_action("Connect")
+        action_connect.setEnabled(False)
+
+        if self.model.system_status["isCrioConnected"]:
+            await prompt_dialog_warning(
+                "_callback_connect()", "There is the connection with M2 controller."
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2gui/main_window.py]: Fixed in the commit: [a7be87c](https://github.com/lsst-ts/ts_m2gui/pull/19/commits/a7be87c8d2aea384541dd1e1fc9a0085be3cc715). Thanks!

File: python/lsst/ts/m2gui/main_window.py [hunk 13]
```diff
@@ -238,62 +265,139 @@ def _add_tool_bar(self):
...
+
+        action_connect.setEnabled(True)
 
     @asyncSlot()
     async def _callback_disconnect(self):
-        """Callback function to disconnect from the M2 controller."""
-        run_command(self.model.disconnect)
+        """Callback function to disconnect from the M2 controller.
+
+        The 'connect', 'disconnect' and 'exit' actions will be disabled first.
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2gui/main_window.py]: Fixed in the commit: [a7be87c](https://github.com/lsst-ts/ts_m2gui/pull/19/commits/a7be87c8d2aea384541dd1e1fc9a0085be3cc715). Thanks!

File: python/lsst/ts/m2gui/model.py [hunk 14]
```diff
@@ -552,18 +607,280 @@ def connect(self):
...
+            if (self.local_mode == LocalMode.Enable) and (
+                self.controller.controller_state == salobj.State.FAULT
+            ):
+                # TODO: Implement this in DM-36015
+                pass
+
+            else:
+                await asyncio.sleep(self.timeout_in_second)
+
+        self.log.debug("Stop the running of event loop from component.")
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2gui/model.py]: Fixed in the commit: [a7be87c](https://github.com/lsst-ts/ts_m2gui/pull/19/commits/a7be87c8d2aea384541dd1e1fc9a0085be3cc715). Thanks!

File: python/lsst/ts/m2gui/model.py [hunk 15]
```diff
@@ -552,18 +607,280 @@ def connect(self):
...
+
+                if (
+                    time_wait_telemetry >= self.TELEMETRY_WAIT_TIMEOUT
+                    and not is_telemetry_timed_out
+                ):
+                    self.log.warning(
+                        (
+                            "No telemetry for more than "
+                            f"{self.TELEMETRY_WAIT_TIMEOUT } seconds "
+                            "already. Is the internet OK?"
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2gui/model.py]: Fixed in the commit: [a7be87c](https://github.com/lsst-ts/ts_m2gui/pull/19/commits/a7be87c8d2aea384541dd1e1fc9a0085be3cc715). Thanks!

File: python/lsst/ts/m2gui/model.py [hunk 16]
```diff
@@ -552,18 +607,280 @@ def connect(self):
...
+                # Send the signals of telemetry
+                self._process_telemetry(message)
+
+            else:
+                self.log.debug(
+                    f"Clients not connected. Waiting {self.timeout_in_second}s..."
+                )
+                await asyncio.sleep(self.timeout_in_second)
+
+        self.log.debug("Telemetry loop from component closed.")
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2gui/model.py]: Fixed in the commit: [a7be87c](https://github.com/lsst-ts/ts_m2gui/pull/19/commits/a7be87c8d2aea384541dd1e1fc9a0085be3cc715). Thanks!

File: python/lsst/ts/m2gui/model.py [hunk 17]
```diff
@@ -552,18 +607,280 @@ def connect(self):
...
+
+        # In the simulation mode, the ports of mock server are randomly
+        # assigned by the operation system.
+        if self._is_simulation_mode:
+            port_command = self._mock_server.server_command.port
+            port_telemetry = self._mock_server.server_telemetry.port
+
+        self.log.debug(f"Host in connection request: {self.host}")
+        self.log.debug(f"Command port in connection request: {port_command}")
+        self.log.debug(f"Telemetry port in connection request: {port_telemetry}")
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2gui/model.py:821]: Fixed in the commit: [a7be87c](https://github.com/lsst-ts/ts_m2gui/pull/19/commits/a7be87c8d2aea384541dd1e1fc9a0085be3cc715). Thanks!

File: python/lsst/ts/m2gui/model.py [hunk 18]
```diff
@@ -552,18 +607,280 @@ def connect(self):
...
+        time_start = time.monotonic()
+        connection_pooling_time = 0.1
+        while not self.controller.are_clients_connected() and (
+            (time.monotonic() - time_start) < self.timeout_connection
+        ):
+            await asyncio.sleep(connection_pooling_time)
+
+        if not self.controller.are_clients_connected():
+            raise RuntimeError(
+                f"Timeount in connection. Host: {self.host}, ports: {port_command} and {port_telemetry}"
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2gui/model.py]: Fixed in the commit: [a7be87c](https://github.com/lsst-ts/ts_m2gui/pull/19/commits/a7be87c8d2aea384541dd1e1fc9a0085be3cc715). Thanks!

File: python/lsst/ts/m2gui/utils.py [hunk 19]
```diff
@@ -251,24 +254,27 @@ def prompt_dialog_warning(title, description, is_prompted=True):
...
+
     if is_prompted:
-        dialog.exec()
+        await dialog.show()
 
 
-def run_command(command, *args, is_prompted=True):
-    """Run the command.
+async def run_command(command, *args, is_prompted=True):
+    """Run the command. This command can be a normal function or coroutine.
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2gui/utils.py]: Fixed in the commit: [a7be87c](https://github.com/lsst-ts/ts_m2gui/pull/19/commits/a7be87c8d2aea384541dd1e1fc9a0085be3cc715). Thanks!

File: python/lsst/ts/m2gui/utils.py [hunk 20]
```diff
@@ -251,24 +254,27 @@ def prompt_dialog_warning(title, description, is_prompted=True):
...
 
     If the command fails, there will be a prompt dialog to warn the users
     by default.
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2gui/utils.py]: Fixed in the commit: [a7be87c](https://github.com/lsst-ts/ts_m2gui/pull/19/commits/a7be87c8d2aea384541dd1e1fc9a0085be3cc715). Thanks!

File: doc/version_history.rst [hunk 21]
```diff
@@ -6,6 +6,15 @@
...
 
+.. _lsst.ts.m2gui-0.1.9:
+
+-------------
+0.1.9
+-------------
+
+* Begin to implement the simulation mode.
+* Support the assignment of debug level.
+
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply doc/version_history.rst:17]: Fixed in the commit: [a7be87c](https://github.com/lsst-ts/ts_m2gui/pull/19/commits/a7be87c8d2aea384541dd1e1fc9a0085be3cc715). Thanks!

**Pull request merged**